ci-tools/fpga-boss/src/main.rs

```diff
@@ -301,10 +346,14 @@ fn main_impl() -> anyhow::Result<()> {
             }
         }
         Some(("serve", sub_matches)) => {
+            // Reuse the previous token if we never connect to GitHub.
+            // This avoids creating a bunch of offline runners if the FPGA fails to establish a
+            // connection.
+            let mut cached_token: Option<Vec<u8>> = None;
             let mut fpga = get_fpga_ftdi()?;
             let mut sd_mux = get_sd_mux()?;
             let sd_dev_path = get_sd_dev_path()?;
-            loop {
+            'outer: loop {
```

Comment:
I wonder if this code would be clearer with a helper function like `run_single_job`. That way you can just have simple error handling in that function and just handle the timeout error condition once in this loop.

Reply:
On the other hand, maybe handling it in both places is better, since it allows for a more specific error message

Reply:
I agree on both counts, the error message isn't super important since it can be inferred from the FPGA logs.

I plan to put the FPGA state management behind a trait, which I think would clean this up

Reply:
The timeout does vary based on what the FPGAs current state is

Reply:
sgtm. Were you planning to do that cleanup in this PR or a follow-up?

Reply:
Whatever is easier to review works for me - this is the code I am using on the host runner, it doesn't have to be checked in.

Reply:
Ok, let's go ahead and check this in and then can clean up in a different PR.